packages/vm/src/instance.rs

```diff
@@ -66,38 +66,38 @@ where
                 //   value region too small: -1_000_001
                 //   key does not exist: -1_001_001
                 // Ownership of both input and output pointer is not transferred to the host.
-                "db_read" => Func::new(move |ctx: &mut Ctx, key_ptr: u32, value_ptr: u32| -> i32 {
+                "db_read" => Func::new(move |ctx: &mut Ctx, key_ptr: u32, value_ptr: u32| -> VmResult<i32> {
                     do_read::<S, Q>(ctx, key_ptr, value_ptr)
                 }),
                 // Writes the given value into the database entry at the given key.
                 // Ownership of both input and output pointer is not transferred to the host.
                 // Returns 0 on success. Returns negative value on error.
-                "db_write" => Func::new(move |ctx: &mut Ctx, key_ptr: u32, value_ptr: u32| -> i32 {
+                "db_write" => Func::new(move |ctx: &mut Ctx, key_ptr: u32, value_ptr: u32| -> VmResult<i32> {
                     do_write::<S, Q>(ctx, key_ptr, value_ptr)
                 }),
                 // Removes the value at the given key. Different than writing &[] as future
                 // scans will not find this key.
                 // At the moment it is not possible to differentiate between a key that existed before and one that did not exist (https://github.com/CosmWasm/cosmwasm/issues/290).
                 // Ownership of both key pointer is not transferred to the host.
                 // Returns 0 on success. Returns negative value on error.
-                "db_remove" => Func::new(move |ctx: &mut Ctx, key_ptr: u32| -> i32 {
+                "db_remove" => Func::new(move |ctx: &mut Ctx, key_ptr: u32| -> VmResult<i32> {
                     do_remove::<S, Q>(ctx, key_ptr)
                 }),
                 // Reads human address from human_ptr and writes canonicalized representation to canonical_ptr.
                 // A prepared and sufficiently large memory Region is expected at canonical_ptr that points to pre-allocated memory.
                 // Returns 0 on success. Returns negative value on error.
                 // Ownership of both input and output pointer is not transferred to the host.
-                "canonicalize_address" => Func::new(move |ctx: &mut Ctx, human_ptr: u32, canonical_ptr: u32| -> i32 {
+                "canonicalize_address" => Func::new(move |ctx: &mut Ctx, human_ptr: u32, canonical_ptr: u32| -> VmResult<i32> {
```

Comment:
Okay, so we expose the rich error value from the host (rust/go) side to the guest (wasm) side. Yes, much nicer than this i32 one. Will we even need the `i32` with the error properly returned? Or would `VmResult<()>` work?

Reply:
You still want the i32 there to be able to report user-triggered errors to the contract. this is done in several places, for example when the contract provides buffers of insufficient length, or incorrect iterator ID. I explained this in some of the resolved comments above